File: mobile/src/main/java/guepardoapps/whosbirthday/controller/DialogController.java
package guepardoapps.whosbirthday.controller;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.support.annotation.NonNull;
import android.view.View;
import android.view.Window;
import android.view.View.OnClickListener;
import android.webkit.CookieManager;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import guepardoapps.whosbirthday.R;
import guepardoapps.whosbirthday.runnable.EditDialogRunnable;
import guepardoapps.whosbirthday.tools.Logger;

public class DialogController {
    private static final String TAG = DialogController.class.getSimpleName();
    private Logger _logger;

    private int _textColor;
    private int _backgroundColor;

    private Context _context;
    private Dialog _dialog;

    public Runnable CloseDialogCallback = new Runnable() {
        @Override
        public void run() {
            if (_dialog != null) {
                _dialog.dismiss();
                _dialog = null;
            }
        }
    };

    public DialogController(Context context, int textColor, int backgroundColor) {
        _logger = new Logger(TAG);
        _logger.Debug("Created new " + TAG + "...");

        _context = context;

        _textColor = textColor;
        _backgroundColor = backgroundColor;

        _logger.Debug("_textColor: " + String.valueOf(_textColor));
        _logger.Debug("_backgroundColor: " + String.valueOf(_backgroundColor));
    }

    public void ShowDialogSingle(
            @NonNull String title,
            @NonNull String prompt,
            @NonNull String ok,
            @NonNull final Runnable okCallback,
            boolean isCancelable) {
        _dialog = new Dialog(_context);

        _dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        _dialog.setContentView(R.layout.custom_dialog_1button);

        LinearLayout backgroundView = _dialog.findViewById(R.id.custom_alert_dialog_background);
        backgroundView.setBackgroundColor(_backgroundColor);

        TextView titleView = _dialog.findViewById(R.id.custom_alert_dialog_title);
        titleView.setText(title);
        titleView.setTextColor(_textColor);

        TextView promptView = _dialog.findViewById(R.id.custom_alert_dialog_prompt);
        promptView.setText(prompt);
        promptView.setTextColor(_textColor);

        Button btnOk = _dialog.findViewById(R.id.custom_alert_dialog_btn_ok);
        btnOk.setText(ok);
        btnOk.setTextColor(_textColor);
        btnOk.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                okCallback.run();
            }
        });

        _dialog.setCancelable(isCancelable);
        _dialog.show();
    }

    public void ShowDialogDouble(
            @NonNull String title,
            @NonNull String prompt,
            @NonNull String ok,
            @NonNull final Runnable okCallback,
            @NonNull String cancel,
            @NonNull final Runnable cancelCallback,
            boolean isCancelable) {
        _dialog = new Dialog(_context);

        _dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        _dialog.setContentView(R.layout.custom_dialog_2buttons);

        LinearLayout backgroundView = _dialog.findViewById(R.id.custom_alert_dialog_background);
        backgroundView.setBackgroundColor(_backgroundColor);

        TextView titleView = _dialog.findViewById(R.id.custom_alert_dialog_title);
        titleView.setText(title);
        titleView.setTextColor(_textColor);

        TextView promptView = _dialog.findViewById(R.id.custom_alert_dialog_prompt);
        promptView.setText(prompt);
        promptView.setTextColor(_textColor);

        Button btnOk = _dialog.findViewById(R.id.custom_alert_dialog_btn_ok);
        btnOk.setText(ok);
        btnOk.setTextColor(_textColor);
        btnOk.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                okCallback.run();
            }
        });

        Button btnCancel = _dialog.findViewById(R.id.custom_alert_dialog_btn_cancel);
        btnCancel.setText(cancel);
        btnCancel.setTextColor(_textColor);
        btnCancel.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                cancelCallback.run();
            }
        });

        _dialog.setCancelable(isCancelable);
        _dialog.show();
    }

    public void ShowDialogTriple(
            @NonNull String title,
            @NonNull String prompt,
            @NonNull String ok,
            @NonNull final Runnable okRunnable,
            @NonNull String neutral,
            @NonNull final Runnable neutralRunnable,
            @NonNull String cancel,
            @NonNull final Runnable cancelRunnable,
            boolean isCancelable) {
        _dialog = new Dialog(_context);

        _dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        _dialog.setContentView(R.layout.custom_dialog_3buttons);

        LinearLayout backgroundView = _dialog.findViewById(R.id.custom_alert_dialog_background);
        backgroundView.setBackgroundColor(_backgroundColor);

        TextView titleView = _dialog.findViewById(R.id.custom_alert_dialog_title);
        titleView.setText(title);
        titleView.setTextColor(_textColor);

        TextView promptView = _dialog.findViewById(R.id.custom_alert_dialog_prompt);
        promptView.setText(prompt);
        promptView.setTextColor(_textColor);

        Button btnOk = _dialog.findViewById(R.id.custom_alert_dialog_btn_ok);
        btnOk.setText(ok);
        btnOk.setTextColor(_textColor);
        btnOk.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                okRunnable.run();
            }
        });

        Button btnNeutral = _dialog.findViewById(R.id.custom_alert_dialog_btn_neutral);
        btnNeutral.setText(neutral);
        btnNeutral.setTextColor(_textColor);
        btnNeutral.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                neutralRunnable.run();
            }
        });

        Button btnCancel = _dialog.findViewById(R.id.custom_alert_dialog_btn_cancel);
        btnCancel.setText(cancel);
        btnCancel.setTextColor(_textColor);
        btnCancel.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                cancelRunnable.run();
            }
        });

        _dialog.setCancelable(isCancelable);
        _dialog.show();
    }

    public void ShowDialogEditText(
            @NonNull String title,
            @NonNull String prompt,
            @NonNull String ok,
            @NonNull final EditDialogRunnable okCallback,
            boolean isCancelable) {
        _dialog = new Dialog(_context);

        _dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        _dialog.setContentView(R.layout.custom_dialog_edittext);

        LinearLayout backgroundView = _dialog.findViewById(R.id.custom_alert_dialog_background);
        backgroundView.setBackgroundColor(_backgroundColor);

        TextView titleView = _dialog.findViewById(R.id.custom_alert_dialog_title);
        titleView.setText(title);
        titleView.setTextColor(_textColor);

        TextView promptView = _dialog.findViewById(R.id.custom_alert_dialog_prompt);
        promptView.setText(prompt);
        promptView.setTextColor(_textColor);

        final EditText inputEditText = _dialog.findViewById(R.id.custom_alert_dialog_edittext);
        inputEditText.setTextColor(_textColor);

        Button btnOk = _dialog.findViewById(R.id.custom_alert_dialog_btn_ok);
        btnOk.setText(ok);
        btnOk.setTextColor(_textColor);
        btnOk.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                String input = inputEditText.getText().toString();

                if (input.length() == 0) {
                    _logger.Error("Input has invalid length 0!");
                    return;
                }

                String[] data = okCallback.GetData();
                if (data.length > 0) {
                    data[0] = input;
                } else {
                    data = new String[]{input};
                }
                okCallback.SetData(data);
                okCallback.run();
            }
        });

        _dialog.setCancelable(isCancelable);
        _dialog.show();
    }

    @SuppressLint("SetJavaScriptEnabled")
    public void ShowDialogWebview(
            @NonNull String title,
            @NonNull String prompt,
            @NonNull String url,
            @NonNull String ok,
            @NonNull final Runnable okCallback,
            boolean isCancelable) {
        _dialog = new Dialog(_context);

        _dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        _dialog.setContentView(R.layout.custom_dialog_webview);

        LinearLayout backgroundView = _dialog.findViewById(R.id.custom_webview_dialog_background);
        backgroundView.setBackgroundColor(_backgroundColor);

        TextView titleView = _dialog.findViewById(R.id.custom_webview_dialog_title);
        titleView.setText(title);
        titleView.setTextColor(_textColor);

        TextView promptView = _dialog.findViewById(R.id.custom_webview_dialog_prompt);
        promptView.setText(prompt);
        promptView.setTextColor(_textColor);

        Button btnOk = _dialog.findViewById(R.id.custom_webview_dialog_btn_ok);
        btnOk.setText(ok);
        btnOk.setTextColor(_textColor);
        btnOk.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                okCallback.run();
            }
        });

        if (!url.startsWith("http://") && !url.startsWith("https://")) {
            _logger.Warn("Fixing url!");
            url = "http://" + url;
        }
        _logger.Debug("Url is: " + url);

        WebView webview = _dialog.findViewById(R.id.custom_webview_dialog_webview);
        webview.getSettings().setUseWideViewPort(true);
        webview.getSettings().setBuiltInZoomControls(true);
        webview.getSettings().setSupportZoom(true);
        webview.getSettings().setJavaScriptEnabled(true);
        webview.getSettings().setLoadWithOverviewMode(true);
        webview.setWebViewClient(new WebViewClient());
        webview.setInitialScale(100);
        CookieManager cookieManager = CookieManager.getInstance();
        cookieManager.setAcceptCookie(false);
        webview.loadUrl(url);

        _dialog.setCancelable(isCancelable);
        _dialog.show();
    }

    public void ShowAlertDialogWebView(
            @NonNull String title,
            @NonNull String url) {
        AlertDialog.Builder alert = new AlertDialog.Builder(_context);
        alert.setTitle(title);

        if (!url.startsWith("http://") && !url.startsWith("https://")) {
            _logger.Warn("Fixing url!");
            url = "http://" + url;
        }
        _logger.Debug("Url is: " + url);

        WebView webView = new WebView(_context);
        webView.setInitialScale(100);
        webView.loadUrl(url);
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                view.loadUrl(url);
                return true;
            }
        });

        alert.setView(webView);
        alert.setNegativeButton("Close", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int id) {
                dialog.dismiss();
            }
        });

        alert.show();
    }
}

File: mobile/src/main/java/guepardoapps/whosbirthday/controller/DialogController.kt
package guepardoapps.whosbirthday.controller

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.Context
import android.view.Window
import android.webkit.CookieManager
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import com.rey.material.app.Dialog
import guepardoapps.whosbirthday.R
import guepardoapps.whosbirthday.R.string.ok
import guepardoapps.whosbirthday.runnable.EditDialogRunnable
import guepardoapps.whosbirthday.utils.Logger

class DialogController(private val context: Context) : IDialogController {
    private val tag: String = DialogController::class.java.simpleName

    private var textColor: Int = 0xFFFFFF
    private var backgroundColor: Int = 0x000000

    override fun setColors(textColor: Int, backgroundColor: Int) {
        this.textColor = textColor
        this.backgroundColor = backgroundColor
    }

    override fun singleButton(title: String, prompt: String,
                              btn1: String, btn1Callback: () -> Unit,
                              isCancelable: Boolean): Dialog {
        val dialog = Dialog(context)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
        dialog.setContentView(R.layout.custom_dialog_1button)

        val backgroundView: LinearLayout = dialog.findViewById(R.id.custom_alert_dialog_background)
        backgroundView.setBackgroundColor(backgroundColor)

        val titleView: TextView = dialog.findViewById(R.id.custom_alert_dialog_title)
        titleView.text = title
        titleView.setTextColor(textColor)

        val promptView: TextView = dialog.findViewById(R.id.custom_alert_dialog_prompt)
        promptView.text = prompt
        promptView.setTextColor(textColor)

        val btnOk: Button = dialog.findViewById(R.id.custom_alert_dialog_btn_ok)
        btnOk.text = btn1
        btnOk.setTextColor(textColor)
        btnOk.setOnClickListener { btn1Callback.invoke() }

        dialog.setCancelable(isCancelable)
        dialog.show()

        return dialog
    }

    override fun doubleButton(title: String, prompt: String,
                              btn1: String, btn1Callback: () -> Unit,
                              btn2: String, btn2Callback: () -> Unit,
                              isCancelable: Boolean): Dialog {
        val dialog = Dialog(context)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
        dialog.setContentView(R.layout.custom_dialog_2buttons)

        val backgroundView: LinearLayout = dialog.findViewById(R.id.custom_alert_dialog_background)
        backgroundView.setBackgroundColor(backgroundColor)

        val titleView: TextView = dialog.findViewById(R.id.custom_alert_dialog_title)
        titleView.text = title
        titleView.setTextColor(textColor)

        val promptView: TextView = dialog.findViewById(R.id.custom_alert_dialog_prompt)
        promptView.text = prompt
        promptView.setTextColor(textColor)

        val btnOk: Button = dialog.findViewById(R.id.custom_alert_dialog_btn_ok)
        btnOk.text = btn1
        btnOk.setTextColor(textColor)
        btnOk.setOnClickListener { btn1Callback.invoke() }

        val btnCancel: Button = dialog.findViewById(R.id.custom_alert_dialog_btn_cancel)
        btnCancel.text = btn2
        btnCancel.setTextColor(textColor)
        btnCancel.setOnClickListener { btn2Callback.invoke() }

        dialog.setCancelable(isCancelable)
        dialog.show()

        return dialog
    }

    override fun tripleButton(title: String, prompt: String,
                              btn1: String, btn1Callback: () -> Unit,
                              btn2: String, btn2Callback: () -> Unit,
                              btn3: String, btn3Callback: () -> Unit,
                              isCancelable: Boolean): Dialog {
        val dialog = Dialog(context)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
        dialog.setContentView(R.layout.custom_dialog_3buttons)

        val backgroundView: LinearLayout = dialog.findViewById(R.id.custom_alert_dialog_background)
        backgroundView.setBackgroundColor(backgroundColor)

        val titleView: TextView = dialog.findViewById(R.id.custom_alert_dialog_title)
        titleView.text = title
        titleView.setTextColor(textColor)

        val promptView: TextView = dialog.findViewById(R.id.custom_alert_dialog_prompt)
        promptView.text = prompt
        promptView.setTextColor(textColor)

        val btnOk: Button = dialog.findViewById(R.id.custom_alert_dialog_btn_ok)
        btnOk.text = btn1
        btnOk.setTextColor(textColor)
        btnOk.setOnClickListener { btn1Callback.invoke() }

        val btnNeutral: Button = dialog.findViewById(R.id.custom_alert_dialog_btn_neutral)
        btnNeutral.text = btn2
        btnNeutral.setTextColor(textColor)
        btnNeutral.setOnClickListener { btn2Callback.invoke() }

        val btnCancel: Button = dialog.findViewById(R.id.custom_alert_dialog_btn_cancel)
        btnCancel.text = btn3
        btnCancel.setTextColor(textColor)
        btnCancel.setOnClickListener { btn3Callback.invoke() }

        dialog.setCancelable(isCancelable)
        dialog.show()

        return dialog
    }

    override fun editText(title: String, prompt: String,
                          btn1: String, btn1Callback: EditDialogRunnable,
                          isCancelable: Boolean): Dialog {
        val dialog = Dialog(context)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
        dialog.setContentView(R.layout.custom_dialog_edittext)

        val backgroundView: LinearLayout = dialog.findViewById(R.id.custom_alert_dialog_background)
        backgroundView.setBackgroundColor(backgroundColor)

        val titleView: TextView = dialog.findViewById(R.id.custom_alert_dialog_title)
        titleView.text = title
        titleView.setTextColor(textColor)

        val promptView: TextView = dialog.findViewById(R.id.custom_alert_dialog_prompt)
        promptView.text = prompt
        promptView.setTextColor(textColor)

        val inputEditText: EditText = dialog.findViewById(R.id.custom_alert_dialog_edittext)
        inputEditText.setTextColor(textColor)

        val btnOk: Button = dialog.findViewById(R.id.custom_alert_dialog_btn_ok)
        btnOk.setText(ok)
        btnOk.setTextColor(textColor)
        btnOk.setOnClickListener {
            val input = inputEditText.text.toString()
            if (input.isEmpty()) {
                Logger.instance.error(tag, "Input has invalid length 0!")
                return@setOnClickListener
            }

            var data = btn1Callback.data
            if (data.isNotEmpty()) {
                data[0] = input
            } else {
                data = arrayOf(input)
            }
            btn1Callback.data = data
            btn1Callback.run()
        }

        dialog.setCancelable(isCancelable)
        dialog.show()

        return dialog
    }

    @SuppressLint("SetJavaScriptEnabled")
    override fun webView(title: String, prompt: String, url: String,
                         btn1: String, btn1Callback: () -> Unit,
                         isCancelable: Boolean): Dialog {
        val dialog = Dialog(context)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
        dialog.setContentView(R.layout.custom_dialog_webview)

        val backgroundView: LinearLayout = dialog.findViewById(R.id.custom_webview_dialog_background)
        backgroundView.setBackgroundColor(backgroundColor)

        val titleView: TextView = dialog.findViewById(R.id.custom_webview_dialog_title)
        titleView.text = title
        titleView.setTextColor(textColor)

        val promptView: TextView = dialog.findViewById(R.id.custom_webview_dialog_prompt)
        promptView.text = prompt
        promptView.setTextColor(textColor)

        val btnOk: Button = dialog.findViewById(R.id.custom_webview_dialog_btn_ok)
        btnOk.text = btn1
        btnOk.setTextColor(textColor)
        btnOk.setOnClickListener { btn1Callback.invoke() }

        var usedUrl = url
        if (!usedUrl.startsWith("http://") && !usedUrl.startsWith("https://")) {
            Logger.instance.warning(tag, "Fixing url!")
            usedUrl = "http://$usedUrl"
        }
        Logger.instance.debug(tag, "UsedUrl is: $usedUrl")

        val webView: WebView = dialog.findViewById(R.id.custom_webview_dialog_webview)
        webView.settings.useWideViewPort = true
        webView.settings.builtInZoomControls = true
        webView.settings.setSupportZoom(true)
        webView.settings.javaScriptEnabled = true
        webView.settings.loadWithOverviewMode = true
        webView.webViewClient = WebViewClient()
        webView.setInitialScale(100)
        val cookieManager = CookieManager.getInstance()
        cookieManager.setAcceptCookie(false)
        webView.loadUrl(url)

        dialog.setCancelable(isCancelable)
        dialog.show()

        return dialog
    }

    @Suppress("OverridingDeprecatedMember")
    override fun alertWebView(title: String, url: String): AlertDialog.Builder {
        val alert = AlertDialog.Builder(context)
        alert.setTitle(title)

        var usedUrl = url
        if (!usedUrl.startsWith("http://") && !usedUrl.startsWith("https://")) {
            Logger.instance.warning(tag, "Fixing url!")
            usedUrl = "http://$usedUrl"
        }
        Logger.instance.debug(tag, "UsedUrl is: $usedUrl")

        val webView = WebView(context)
        webView.setInitialScale(100)
        webView.loadUrl(usedUrl)
        webView.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
                view.loadUrl(url)
                return true
            }
        }

        alert.setView(webView)
        alert.setNegativeButton("Close") { dialog, _ -> dialog.dismiss() }
        alert.show()

        return alert
    }
}